### PR TITLE
Updated nuget references and saga code to Slq Persistence

### PR DESCRIPTION
**Pull request by seanfarmar**
Moving from NHibernate persistence to SQL Persistence

### CODE REVIEW

**Comment by seanfarmar** [comment]
Still need to update the documentation

File: .gitignore [hunk 1]
```diff
@@ -24,4 +24,5 @@ Desktop.ini
 $RECYCLE.BIN/
 
 # Mac desktop service store files
-.DS_Store
\ No newline at end of file
+.DS_Store
+/exercises/02-publish-subscribe/after/Divergent.Customers/Divergent.Customers.csproj.user
```
Comment (adamralph) [inline .gitignore]: This change should be reverted, or a line should be added to ignore _all_ `*.csproj.user` files (should be in a separate commit/PR).

File: exercises/01-composite-ui/before/Divergent.Sales.API/Divergent.Sales.API.csproj [hunk 2]
```diff
@@ -9,6 +9,10 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNet.WebApi.Cors" Version="5.2.3" />
     <PackageReference Include="Microsoft.AspNet.WebApi.OwinSelfHost" Version="5.2.4" />
+    <PackageReference Include="NServiceBus" Version="7.0.1" />
+    <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="2.1.0" />
+    <PackageReference Include="NServiceBus.Transport.Msmq" Version="1.0.0" />
+    <PackageReference Include="Radical.Bootstrapper.Windsor.AspNet" Version="1.0.0" />
```
Comment (adamralph) [inline exercises/01-composite-ui/before/Divergent.Sales.API/Divergent.Sales.API.csproj]: Exercise 1 should not have any NServiceBus references. The changes to this file should be reverted.

File: exercises/01-composite-ui/before/Divergent.Sales.ViewModelComposition/Divergent.Sales.ViewModelComposition.csproj [hunk 3]
```diff
@@ -5,9 +5,9 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Divergent.Sales.ViewModelComposition.Events\Divergent.Sales.ViewModelComposition.Events.csproj" />
     <ProjectReference Include="..\..\..\shared-api-gateway\ITOps.Json\ITOps.Json.csproj" />
     <ProjectReference Include="..\..\..\shared-api-gateway\ITOps.ViewModelComposition\ITOps.ViewModelComposition.csproj" />
-    <ProjectReference Include="..\Divergent.Sales.ViewModelComposition.Events\Divergent.Sales.ViewModelComposition.Events.csproj" />
```
Comment (adamralph) [inline exercises/01-composite-ui/before/Divergent.Sales.ViewModelComposition/Divergent.Sales.ViewModelComposition.csproj]: Changes to this file should be reverted.

File: exercises/02-publish-subscribe/after/Divergent.Customers/Divergent.Customers.csproj [hunk 4]
```diff
@@ -7,6 +7,9 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="NServiceBus" Version="7.0.1" />
+    <PackageReference Include="NServiceBus.Persistence.Sql" Version="4.2.0" />
+    <PackageReference Include="NServiceBus.Persistence.Sql.MsBuild" Version="4.2.0" />
```
Comment (adamralph) [inline exercises/02-publish-subscribe/after/Divergent.Customers/Divergent.Customers.csproj]: These changes should not be made here. It is the shared `ITOps.EndpointConfiguration' project that defines the transport. Change should be made there: https://github.com/Particular/Workshop/blob/6aee55e19b8d85f431f76eaa9971176f2c2eff63/exercises/shared-api-gateway/ITOps.EndpointConfig/ITOps.EndpointConfig.csproj#L8

The same applies to all the other places where these references have been added.

File: exercises/02-publish-subscribe/after/Divergent.Finance/Handlers/OrderSubmittedHandler.cs [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
 using System.Threading.Tasks;
-using Divergent.Finance.Data.Context;
-using Divergent.Finance.Data.Models;
-using Divergent.Finance.Messages.Commands;
 using Divergent.Sales.Messages.Events;
 using NServiceBus;
 using NServiceBus.Logging;
+using Divergent.Finance.Data.Context;
+using Divergent.Finance.Data.Models;
+using Divergent.Finance.Messages.Commands;
```
Comment (adamralph) [inline exercises/02-publish-subscribe/after/Divergent.Finance/Handlers/OrderSubmittedHandler.cs]: Should be reverted.

File: exercises/03-sagas/after/Divergent.Customers/Divergent.Customers.csproj.user [hunk 6]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <ShowAllFiles>true</ShowAllFiles>
+  </PropertyGroup>
+</Project>
```
Comment (adamralph) [inline exercises/03-sagas/after/Divergent.Customers/Divergent.Customers.csproj.user]: Should be reverted.

File: exercises/03-sagas/after/Divergent.Shipping/Sagas/ShippingSaga.cs [hunk 7]
```diff
@@ -7,16 +7,20 @@
 
 namespace Divergent.Shipping.Sagas
 {
-    class ShippingSaga : Saga<ShippingSagaData>,
+    using NServiceBus.Persistence.Sql;
+
+    class ShippingSaga : SqlSaga<ShippingSagaData>,
```
Comment (adamralph) [inline exercises/03-sagas/after/Divergent.Shipping/Sagas/ShippingSaga.cs]: This is the change that we wanted to avoid. Does the new version of SQLP not support standard core sagas?

File: exercises/03-sagas/after/Divergent.Shipping/Sagas/ShippingSagaData.cs [hunk 8]
```diff
@@ -5,17 +5,17 @@ namespace Divergent.Shipping.Sagas
...
+        public int OrderId { get; set; }
+        public int CustomerId { get; set; }
 
-        public virtual bool IsPaymentProcessed { get; set; }
-        public virtual bool IsOrderSubmitted { get; set; }
-
-        public virtual ICollection<Product> Products { get; set; }
+        public bool IsOrderSubmitted { get; set; }
+        public bool IsPaymentProcessed { get; set; }
+
```
Comment (adamralph) [inline exercises/03-sagas/after/Divergent.Shipping/Sagas/ShippingSagaData.cs]: This line should be reverted.

**Comment by dvdstelt** [comment]
I think these are the changes we need. Removed some NServiceBus NuGet references that were also already removed in other solutions.

**Comment by dvdstelt** [comment]
I ran every exercise and they seem to be working. To be 100% sure I need to do exercise 03 myself to verify if it is still correct.

File: .gitignore [hunk 9]
```diff
@@ -24,4 +24,5 @@ Desktop.ini
 $RECYCLE.BIN/
 
 # Mac desktop service store files
-.DS_Store
\ No newline at end of file
+.DS_Store
+*.csproj.user
```
Comment (adamralph) [inline .gitignore]: I'd prefer to have this in a separate commit.

File: exercises/02-publish-subscribe/after/Divergent.Finance/Divergent.Finance.csproj [hunk 10]
```diff
@@ -8,7 +8,6 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNet.WebApi.Client" Version="5.2.3" />
-    <PackageReference Include="NServiceBus" Version="7.0.1" />
```
Comment (adamralph) [inline exercises/02-publish-subscribe/after/Divergent.Finance/Divergent.Finance.csproj]: I'd prefer to have this, and the other similar removals, in a separate commit.

File: exercises/03-sagas/after/Divergent.Shipping/Divergent.Shipping.csproj [hunk 11]
```diff
@@ -7,7 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="NServiceBus" Version="7.0.1" />
+    <PackageReference Include="NServiceBus.Persistence.Sql.MsBuild" Version="4.2.0" />
```
Comment (adamralph) [inline exercises/03-sagas/after/Divergent.Shipping/Divergent.Shipping.csproj]: Is this reference required in this project (and in other similar places)?

File: exercises/03-sagas/after/Divergent.Shipping/Sagas/ShippingSagaData.cs [hunk 12]
```diff
@@ -5,17 +5,15 @@ namespace Divergent.Shipping.Sagas
...
-
-        public virtual bool IsPaymentProcessed { get; set; }
-        public virtual bool IsOrderSubmitted { get; set; }
-
-        public virtual ICollection<Product> Products { get; set; }
+        public int OrderId { get; set; }
+        public int CustomerId { get; set; }
+        public bool IsOrderSubmitted { get; set; }
+        public bool IsPaymentProcessed { get; set; }
+        public ICollection<Product> Products { get; set; }
```
Comment (adamralph) [inline exercises/03-sagas/after/Divergent.Shipping/Sagas/ShippingSagaData.cs]: I'd prefer to have the whitespace alteration and re-ordering of properties (and in other similar places) in a separate commit. They are unrelated changes, and also make it more difficult to spot a typo.

This is a minor point though, and I wouldn't block the PR on it.

File: exercises/04-integration/after/Divergent.Shipping/Sagas/ShippingSaga.cs [hunk 13]
```diff
@@ -8,6 +8,8 @@
 
 namespace Divergent.Shipping.Sagas
 {
+    using NServiceBus.Persistence.Sql;
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Shipping/Sagas/ShippingSaga.cs]: This is redundant and should be reverted.

File: exercises/04-integration/after/Divergent.Shipping/Sagas/ShippingSagaData.cs [hunk 14]
```diff
@@ -3,21 +3,17 @@
 
 namespace Divergent.Shipping.Sagas
 {
-    public class ShippingSagaData : ContainSagaData
+    class ShippingSagaData : ContainSagaData
```
Comment (adamralph) [inline exercises/04-integration/after/Divergent.Shipping/Sagas/ShippingSagaData.cs]: Why was the visibility changed?

File: exercises/04-integration/before/Divergent.Shipping/Sagas/ShippingSagaData.cs [hunk 15]
```diff
@@ -6,12 +6,9 @@ namespace Divergent.Shipping.Sagas
     public class ShippingSagaData : ContainSagaData
     {
         public virtual int OrderId { get; set; }
-
```
Comment (adamralph) [inline exercises/04-integration/before/Divergent.Shipping/Sagas/ShippingSagaData.cs]: Should these properties also be made non-virtual?

File: exercises/shared-api-gateway/ITOps.EndpointConfig/EndpointConfigurationExtensions.cs [hunk 16]
```diff
@@ -1,11 +1,13 @@
 using NServiceBus;
 using NServiceBus.Logging;
-using NServiceBus.Persistence;
 using System;
 using System.IO;
 
 namespace ITOps.EndpointConfig
 {
+    using System.Data.SqlClient;
+    using NServiceBus.Persistence.Sql;
```
Comment (adamralph) [inline exercises/shared-api-gateway/ITOps.EndpointConfig/EndpointConfigurationExtensions.cs]: Should be outside the namespace.

File: exercises/shared-api-gateway/ITOps.EndpointConfig/ITOps.EndpointConfig.csproj [hunk 17]
```diff
@@ -5,7 +5,8 @@
   <ItemGroup>
     <PackageReference Include="NServiceBus" Version="7.0.1" />
     <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="2.1.0" />
-    <PackageReference Include="NServiceBus.NHibernate" Version="8.0.0" />
+    <PackageReference Include="NServiceBus.Persistence.Sql" Version="4.2.0" />
+    <PackageReference Include="NServiceBus.Persistence.Sql.MsBuild" Version="4.2.0" />
```
Comment (adamralph) [inline exercises/shared-api-gateway/ITOps.EndpointConfig/ITOps.EndpointConfig.csproj]: Does this reference have to be in this project as well as the endpoint projects themselves?

Comment (dvdstelt) [reply exercises/shared-api-gateway/ITOps.EndpointConfig/ITOps.EndpointConfig.csproj]: Only in Shipping, or otherwise the saga sql for the table isn't generated.

Comment (adamralph) [reply exercises/shared-api-gateway/ITOps.EndpointConfig/ITOps.EndpointConfig.csproj]: OK, so we can remove it from this project?

File: exercises/03-sagas/after/Divergent.Shipping/Divergent.Shipping.csproj (hunk 11 above)
Comment (seanfarmar) [reply exercises/03-sagas/after/Divergent.Shipping/Divergent.Shipping.csproj]: It seems we need to have them to get the build to work

File: exercises/04-integration/after/Divergent.Shipping/Sagas/ShippingSagaData.cs (hunk 14 above)
Comment (seanfarmar) [reply exercises/04-integration/after/Divergent.Shipping/Sagas/ShippingSagaData.cs]: Fixed in next push

**Comment by seanfarmar** [comment]
pushed amendments

**Comment by seanfarmar** [comment]
@adamralph 
>I'd still prefer to have the unrelated changes in a separate commit but they are relatively minor so I'm willing overlook it.

If you want to get on a call tomorrow and walk me through this I'm happy to ammend it :-)

**Comment by adamralph** [comment]
@seanfarmar @dvdstelt I've scheduled a time tomorrow morning for us to do the remaning git-fu.

File: exercises/shared-api-gateway/ITOps.EndpointConfig/EndpointConfigurationExtensions.cs [hunk 18]
```diff
@@ -1,8 +1,9 @@
-using NServiceBus;
-using NServiceBus.Logging;
-using NServiceBus.Persistence;
 using System;
 using System.IO;
+using System.Data.SqlClient;
+using NServiceBus.Persistence.Sql;
+using NServiceBus;
+using NServiceBus.Logging;
```
Comment (adamralph) [inline exercises/shared-api-gateway/ITOps.EndpointConfig/EndpointConfigurationExtensions.cs]: These usings need sorting (mentioned to @dvdstelt and he will fix).

**Comment by dvdstelt** [comment]
Apparently, many more references were needed because of subscription storage tables that needed to be generated.

File: exercises/03-sagas/before/Divergent.Customers/Divergent.Customers.csproj [hunk 19]
```diff
@@ -7,6 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="NServiceBus.Persistence.Sql.MsBuild" Version="4.2.0" />
```
Comment (adamralph) [inline exercises/03-sagas/before/Divergent.Customers/Divergent.Customers.csproj]: This change has not been made in the after project.

File: exercises/03-sagas/before/Divergent.Finance/Divergent.Finance.csproj [hunk 20]
```diff
@@ -8,6 +8,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNet.WebApi.Client" Version="5.2.3" />
+    <PackageReference Include="NServiceBus.Persistence.Sql.MsBuild" Version="4.2.0" />
```
Comment (adamralph) [inline exercises/03-sagas/before/Divergent.Finance/Divergent.Finance.csproj]: This change has not been made in the after project.

File: exercises/03-sagas/before/Divergent.Sales/Divergent.Sales.csproj [hunk 21]
```diff
@@ -7,6 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="NServiceBus.Persistence.Sql.MsBuild" Version="4.2.0" />
```
Comment (adamralph) [inline exercises/03-sagas/before/Divergent.Sales/Divergent.Sales.csproj]: This change has not been made in the after project.

File: exercises/04-integration/before/Divergent.ITOps/Divergent.ITOps.csproj [hunk 22]
```diff
@@ -9,8 +9,8 @@
   <ItemGroup>
     <PackageReference Include="Autofac" Version="4.6.1" />
     <PackageReference Include="EntityFramework" Version="6.1.3" />
-    <PackageReference Include="NServiceBus" Version="7.0.1" />
```
Comment (adamralph) [inline exercises/04-integration/before/Divergent.ITOps/Divergent.ITOps.csproj]: This change has not been made in the after project. Also, this change should be in a separate PR, since it's unrelated to the sqlp switch.

**Comment by adamralph** [comment]
Oh, and I raised https://github.com/Particular/Workshop/pull/287 to replace https://github.com/Particular/Workshop/commit/0eaa509bc1c64afc1bd6f7a862cd9ee7b0ac22ae
